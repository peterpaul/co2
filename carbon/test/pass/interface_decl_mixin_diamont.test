["stdio.h"] int printf(char * format, ...);

interface Foo {
	int foo () {
		return 2;
	}
}

interface Bar, Foo {
	int bar () {
		return foo () + 1;
	}
}

interface Baz, Foo {
	int baz () {
		return foo () + 2;
	}
}

class FooBar, Bar, Baz {
}

int main () {
	FooBar fooBar = new FooBar ();

	Foo foo = (Foo) fooBar;
	Bar bar = (Bar) fooBar;
	Baz baz = (Baz) fooBar;

	printf ("foo.foo () = %d\n", foo.foo ());
	printf ("bar.foo () = %d\n", bar.foo ());
	printf ("bar.bar () = %d\n", bar.bar ());
	printf ("baz.foo () = %d\n", baz.foo ());
	printf ("baz.baz () = %d\n", baz.baz ());

	printf ("fooBar.foo () = %d\n", fooBar.foo ());
	printf ("fooBar.foo () = %d\n", fooBar.foo ());
	printf ("fooBar.bar () = %d\n", fooBar.bar ());
	printf ("fooBar.foo () = %d\n", fooBar.foo ());
	printf ("fooBar.baz () = %d\n", fooBar.baz ());

	return 0;
}
