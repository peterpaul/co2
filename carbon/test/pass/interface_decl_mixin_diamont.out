foo.foo () = 2
bar.foo () = 2
bar.bar () = 3
baz.foo () = 2
baz.baz () = 4
fooBar.foo () = 2
fooBar.foo () = 2
fooBar.bar () = 3
fooBar.foo () = 2
fooBar.baz () = 4
